rpc/src/v1/impls/signing.rs

```diff
@@ -45,61 +46,76 @@ use v1::types::{
 	Origin,
 };
 
+use parity_reactor::Remote;
+
 /// After 60s entries that are not queried with `check_request` will get garbage collected.
 const MAX_PENDING_DURATION_SEC: u32 = 60;
-/// Max number of total requests pending and completed, before we start garbage collecting them.
-const MAX_TOTAL_REQUESTS: usize = SIGNING_QUEUE_LIMIT;
 
+#[must_use = "futures do nothing unless polled"]
 enum DispatchResult {
-	Promise(ConfirmationPromise),
+	Future(RpcConfirmationReceiver),
 	Value(RpcConfirmationResponse),
 }
 
-/// Implementation of functions that require signing when no trusted signer is used.
-pub struct SigningQueueClient<D> {
-	signer: Arc<SignerService>,
-	accounts: Option<Arc<AccountProvider>>,
-	dispatcher: D,
-	pending: Arc<Mutex<TransientHashMap<U256, ConfirmationPromise>>>,
+impl Future for DispatchResult {
+	type Item = RpcConfirmationResponse;
+	type Error = Error;
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		match *self {
+			DispatchResult::Value(ref response) => Ok(Async::Ready(response.clone())),
+			DispatchResult::Future(ref mut future) => match try_ready!(future.poll()) {
+				RpcConfirmationOutcome::Rejected => Err(errors::request_rejected()),
+				RpcConfirmationOutcome::Confirmed(response) => Ok(Async::Ready(response)),
+				RpcConfirmationOutcome::Failed(error) => Err(error)
+			}
+		}
+	}
 }
 
-fn handle_dispatch<OnResponse>(res: Result<DispatchResult, Error>, on_response: OnResponse)
-	where OnResponse: FnOnce(Result<RpcConfirmationResponse, Error>) + Send + 'static
-{
-	match res {
-		Ok(DispatchResult::Value(result)) => on_response(Ok(result)),
-		Ok(DispatchResult::Promise(promise)) => {
-			promise.wait_for_result(move |result| {
-				on_response(result.unwrap_or_else(|| Err(errors::request_rejected())))
-			})
-		},
-		Err(e) => on_response(Err(e)),
+fn schedule(remote: Remote, pending: Arc<RwLock<BTreeSet<U256>>>, confirmations: Arc<Mutex<TransientHashMap<U256, RpcConfirmationOutcome>>>, future: RpcConfirmationReceiver) {
+	let id = future.id();
+	{
+		let mut pending = pending.write();
+		pending.insert(id.clone());
 	}
-}
 
-fn collect_garbage(map: &mut TransientHashMap<U256, ConfirmationPromise>) {
-	map.prune();
-	if map.len() > MAX_TOTAL_REQUESTS {
-		// Remove all non-waiting entries.
-		let non_waiting: Vec<_> = map
-			.iter()
-			.filter(|&(_, val)| val.result() != ConfirmationResult::Waiting)
-			.map(|(key, _)| *key)
-			.collect();
-		for k in non_waiting {
-			map.remove(&k);
+	let future = future.then(move |result| {
+		{
+			let mut pending = pending.write();
```

Comment:
Seems that this introduces a race condition when the feature is no longer pending but the result is not in confirmations.
Why do we need  two structures at all? Why not `HashMap<U256, Status>`, where `enum Status { Pending, Done(Result..) }`?

Reply:
@tomusdrw Well, my idea was having a LRU-cache (`TransientHashMap`) of successfull confirmations, and a usual Set of UIDs of pending ones. This way we won't have to override the TransientHashMap behavior to collect obsolete confirmations.

My original line of thinking was that get-and-set from a HashMap is not atomic, so the worst case is the same as with two structures; but I think I ignored the fact that we don't need to actualy _read_ from the HashMap to update it's value (we have all the data in the callback, so just overwriting the value will do).

But considering all that, I have a question — the old code had that garbage collection routine (which was actually just touching all the items on the HashMap in the iteration order, sorta invalidationg the point of LRU cache.)

How should I better address it when I put both Pending and Confirmed transactions in the LRU?
I guess I would need a more specialized structure then, which would evict only Confirmed transactions?
Othewise we might end up kicking out some Pending transactions from the queue, and that will make RPC seem unreliable.

Reply:
> My original line of thinking was that get-and-set from a HashMap is not atomic,
It's under a `Mutex` anyway.

> the old code had that garbage collection routine
So the idea with `TransientHashMap` is that elements are evicted after some timeout if they are not actively queried for. It's not LRU and I don't think we actually need it here.
In case some pending transactions are kicked out it means that no-one really cares about them any more (since he is not asking for the status using `checkRequest`).